<a href="https://colab.research.google.com/github/tuchanski/spambase-classification/blob/main/main.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Final Report - Data Science

## Bachelor's Degree in Computer Science / PUCPR (2026-2)

**Prof. Rayson Laroca**

`Guilherme Tuchanski Rocha` - `guilherme.tuchanski@gmail.com | guilherme.tuchanski@pucpr.edu.br`

### Project Overview: Email Spam Classification

This project addresses the task of **email spam classification** using the **Spambase** dataset from the UCI Machine Learning Repository. The objective is to analyze characteristics extracted from email messages and investigate how these features differ between spam and legitimate emails.

The dataset contains numerical features related to the frequency of specific words and characters, as well as information about sequences of consecutive capital letters. These characteristics can be explored to identify patterns associated with spam messages.

During the project, the dataset will be analyzed through descriptive statistics, univariate and multivariate analyses, and data visualizations. In later stages, machine learning models will be trained and evaluated to classify emails as spam or non-spam.

#### Dataset Characteristics

* **Instances:** 4,601 records in the original dataset.
* **Features:** 57 numerical features describing word frequencies, character frequencies, and capitalization patterns.
* **Target:** The `Class` variable, where `1` represents spam and `0` represents non-spam.
* **Task:** Binary classification.
* **Source:** UCI Machine Learning Repository.

# Import the libs you need

In [1]:
!pip install ucimlrepo -q

In [2]:
# Import all packages you need here
import pandas as pd
import numpy as np
from ucimlrepo import fetch_ucirepo

# Add others as needed (seaborn, sklearn, etc.)

# Load your data

Below, load the data using pandas and perform all necessary data cleaning to ensure that the data types are correct for further analysis.

In [3]:
# Load the original data and keep a copy for comparison after cleaning.
spambase = fetch_ucirepo(id=94)

X = spambase.data.features
y = spambase.data.targets

df_raw = pd.concat([X, y], axis=1).copy()
df = df_raw.copy()
df.head()

,word_freq_make,word_freq_address,word_freq_all,word_freq_3d,word_freq_our,word_freq_over,word_freq_remove,word_freq_internet,word_freq_order,word_freq_mail,...,char_freq_;,char_freq_(,char_freq_[,char_freq_!,char_freq_$,char_freq_#,capital_run_length_average,capital_run_length_longest,capital_run_length_total,Class
0,0.00,0.64,0.64,0.0,0.32,0.00,0.00,0.00,0.00,0.00,...,0.00,0.000,0.0,0.778,0.000,0.000,3.756,61,278,1
1,0.21,0.28,0.50,0.0,0.14,0.28,0.21,0.07,0.00,0.94,...,0.00,0.132,0.0,0.372,0.180,0.048,5.114,101,1028,1
2,0.06,0.00,0.71,0.0,1.23,0.19,0.19,0.12,0.64,0.25,...,0.01,0.143,0.0,0.276,0.184,0.010,9.821,485,2259,1
3,0.00,0.00,0.00,0.0,0.63,0.00,0.31,0.63,0.31,0.63,...,0.00,0.137,0.0,0.137,0.000,0.000,3.537,40,191,1
4,0.00,0.00,0.00,0.0,0.63,0.00,0.31,0.63,0.31,0.63,...,0.00,0.135,0.0,0.135,0.000,0.000,3.537,40,191,1


In [4]:
print(df.shape)
print(df.info())
print(df.isnull().sum().sum())

df.head()

(4601, 58)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report      

,word_freq_make,word_freq_address,word_freq_all,word_freq_3d,word_freq_our,word_freq_over,word_freq_remove,word_freq_internet,word_freq_order,word_freq_mail,...,char_freq_;,char_freq_(,char_freq_[,char_freq_!,char_freq_$,char_freq_#,capital_run_length_average,capital_run_length_longest,capital_run_length_total,Class
0,0.00,0.64,0.64,0.0,0.32,0.00,0.00,0.00,0.00,0.00,...,0.00,0.000,0.0,0.778,0.000,0.000,3.756,61,278,1
1,0.21,0.28,0.50,0.0,0.14,0.28,0.21,0.07,0.00,0.94,...,0.00,0.132,0.0,0.372,0.180,0.048,5.114,101,1028,1
2,0.06,0.00,0.71,0.0,1.23,0.19,0.19,0.12,0.64,0.25,...,0.01,0.143,0.0,0.276,0.184,0.010,9.821,485,2259,1
3,0.00,0.00,0.00,0.0,0.63,0.00,0.31,0.63,0.31,0.63,...,0.00,0.137,0.0,0.137,0.000,0.000,3.537,40,191,1
4,0.00,0.00,0.00,0.0,0.63,0.00,0.31,0.63,0.31,0.63,...,0.00,0.135,0.0,0.135,0.000,0.000,3.537,40,191,1


In [5]:
# Start from the original data so this section can be run again safely.
df = df_raw.copy()
print(f"Number of records before cleaning: {len(df)}")

Number of records before cleaning: 4601


In [6]:
# 1. Require numeric values; stop if an unexpected text value is found.
df = df.apply(pd.to_numeric, errors="raise")

# Missing or infinite values need investigation before analysis can continue.
missing_values = df.isna().sum().sum()
infinite_values = np.isinf(df.to_numpy()).sum()
print(f"Missing values: {missing_values}")
print(f"Infinite values: {infinite_values}")
assert missing_values == 0, "Investigate missing values before continuing."
assert infinite_values == 0, "Investigate infinite values before continuing."

Missing values: 0
Infinite values: 0


In [7]:
# 2. Validate labels BEFORE converting to integers (0.5 must not become 0).
assert df["Class"].isin([0, 1]).all(), "Class must contain only 0 or 1."
df["Class"] = df["Class"].astype(int)
print(f"Class values: {sorted(df['Class'].unique())}")

Class values: [np.int64(0), np.int64(1)]


In [8]:
# 3. Frequencies are percentages, so valid values range from 0 to 100.
frequency_cols = df.filter(regex=r"^(word_freq_|char_freq_)").columns
assert df[frequency_cols].ge(0).all().all(), "Negative frequency found."
assert df[frequency_cols].le(100).all().all(), "Frequency above 100 found."

# Zero frequencies mean absence; keep them and do not remove outliers blindly.
average = df["capital_run_length_average"]
longest = df["capital_run_length_longest"]
total = df["capital_run_length_total"]
assert average.ge(0).all(), "Negative capital run length found."
assert (average <= longest).all(), "Average run length exceeds the longest run."
assert (longest <= total).all(), "Longest run exceeds total capital letters."
assert ((longest % 1 == 0) & (total % 1 == 0)).all(), "Capital counts must be integers."

In [9]:
# 4. Keep one copy of each identical row to reduce repeated observations.
# This is an analysis choice: matching features do not prove emails are identical.
duplicates = df.duplicated().sum()
df = df.drop_duplicates().copy()
print(f"Duplicate rows removed: {duplicates}")

# Check for identical features with different labels; keep them for review.
feature_cols = df.columns.drop("Class").tolist()
label_counts = df.groupby(feature_cols, dropna=False)["Class"].nunique()
print(f"Feature groups with conflicting labels: {(label_counts > 1).sum()}")
# Later, keep identical feature groups in the same train/test partition.

Duplicate rows removed: 391
Feature groups with conflicting labels: 3


In [10]:
# 5. Compare class percentages to see the effect of removing duplicates.
class_balance = pd.DataFrame({
    "Before (%)": df_raw["Class"].value_counts(normalize=True) * 100,
    "After (%)": df["Class"].value_counts(normalize=True) * 100,
}).sort_index()
display(class_balance.round(2))

,Before (%),After (%)
Class,,
0,60.6,60.12
1,39.4,39.88


In [11]:
# 6. Rebuild X and y so later analysis uses the cleaned data.
X = df.drop(columns="Class").copy()
y = df["Class"].copy()

# Fit scaling or imputation only on training data during the modeling stage.
print("Data types:")
print(df.dtypes.value_counts())

Data types:
float64    55
int64       3
Name: count, dtype: int64


In [12]:
# Final summary; retain original row indices to trace records back to df_raw.
print(f"Number of records after cleaning: {len(df)}")
print(f"Number of features: {X.shape[1]}")
display(df.head())

Number of records after cleaning: 4210
Number of features: 57


,word_freq_make,word_freq_address,word_freq_all,word_freq_3d,word_freq_our,word_freq_over,word_freq_remove,word_freq_internet,word_freq_order,word_freq_mail,...,char_freq_;,char_freq_(,char_freq_[,char_freq_!,char_freq_$,char_freq_#,capital_run_length_average,capital_run_length_longest,capital_run_length_total,Class
0,0.00,0.64,0.64,0.0,0.32,0.00,0.00,0.00,0.00,0.00,...,0.00,0.000,0.0,0.778,0.000,0.000,3.756,61,278,1
1,0.21,0.28,0.50,0.0,0.14,0.28,0.21,0.07,0.00,0.94,...,0.00,0.132,0.0,0.372,0.180,0.048,5.114,101,1028,1
2,0.06,0.00,0.71,0.0,1.23,0.19,0.19,0.12,0.64,0.25,...,0.01,0.143,0.0,0.276,0.184,0.010,9.821,485,2259,1
3,0.00,0.00,0.00,0.0,0.63,0.00,0.31,0.63,0.31,0.63,...,0.00,0.137,0.0,0.137,0.000,0.000,3.537,40,191,1
4,0.00,0.00,0.00,0.0,0.63,0.00,0.31,0.63,0.31,0.63,...,0.00,0.135,0.0,0.135,0.000,0.000,3.537,40,191,1


# Statistical Description

In this section, you should report the key characteristics of the dataset, including but not limited to:
* the number of instances;
* the number of features;
* the type of target variable or task;
* the number of classes, for classification problems;
* the class distribution, for classification problems;
* the distribution and range of the target variable, for regression problems.

In [16]:
features = df.drop(columns="Class")
target = df["Class"]

summary = pd.Series({
    "Number of instances": len(df),
    "Number of features": features.shape[1],
    "Feature type": "Numeric (word/character frequencies and capital run lengths)",
    "Task": "Binary classification",
    "Target variable": "Class (0 = non-spam, 1 = spam)",
    "Number of classes": target.nunique(),
}, name="Value")
display(summary.to_frame())

,Value
Number of instances,4210
Number of features,57
Feature type,Numeric (word/character frequencies and capita...
Task,Binary classification
Target variable,"Class (0 = non-spam, 1 = spam)"
Number of classes,2


In [17]:
# Show both counts and percentages to make the class balance easy to compare.
class_distribution = target.value_counts().sort_index().rename("Count").to_frame()
class_distribution["Percentage (%)"] = (class_distribution["Count"] / len(df) * 100).round(2)
class_distribution.index = class_distribution.index.map({0: "Non-spam (0)", 1: "Spam (1)"})
class_distribution.index.name = "Class"
display(class_distribution)

,Count,Percentage (%)
Class,,
Non-spam (0),2531,60.12
Spam (1),1679,39.88


In [18]:
# Use the majority class as a simple reference for later model evaluation.
majority_percentage = class_distribution["Count"].max() / len(df) * 100
print(f"Always predicting the majority class would give {majority_percentage:.2f}% accuracy.")

Always predicting the majority class would give 60.12% accuracy.


The target is categorical even though it is encoded as integers: **0 means non-spam** and **1 means spam**. The 57 predictors describe word frequencies (48 features), character frequencies (6 features), and capital letter sequences (3 features).

After cleaning, the dataset contains **4,210 emails**: **2,531 non-spam (60.12%)** and **1,679 spam (39.88%)**. Non-spam is the majority class, but both classes are well represented. Accuracy should be compared with the majority-class baseline and complemented by precision, recall, and F1-score during model evaluation. These percentages describe this dataset and should not be assumed to represent all email traffic.

# Univariate data analysis

In this section, you should perform univariate data analysis on at least **20 variables**.

In the end, you should describe the main variables that are of your interest, and these should be accounted for in the next sections of the report.
The definition of each variable chosen should be clarified, so arbitrary selections are **not** accepted at this point.

For each variable plotted, make sure you determine the following:
1. The distribution of the data (Gaussian, binomial, exponential, etc.);
2. Skewness;
3. Kurtosis;
4. Mean, standard deviation, and what they stand for in the context of the dataset.

Ensure that each variable is **plotted correctly** based on its type. For instance, make sure scatterplots are not used for categorical data and so forth.

In [ ]:
# place as many cells to plot the visualizations,
# as well as to describe the main findings

In [ ]:
# if you realize you need to further clean your data here, there is no problem,
# yet, make sure you are describing the entire process and the rationale
# behind your choices

# Multivariate data analysis

In this section, you should plot at least **5 multivariate visualizations**. The key here is to investigate underlying correlations and behaviors within the dataset.
Naturally, as visualizations are being created, we should end up with obvious results, yet, you should find at least **ONE** non-obvious behavior in the data.

Please follow these steps for creating your visualizations:
1. State an hypothesis. Explain why you have selected these specific variables and what you expect to discover through their relationship;
2. Determine what kind of visualization is the most suitable;
3. Report the findings and discuss whether they corroborate or not the aforementioned hypothesis.


**Hints**

In this section, make sure you go beyond naive explorations. For example, consider applying techniques such as PCA, t-SNE, or even others that we haven't covered in the lectures. The goal is to cultivate a critical mindset toward data analysis and our work.

**Important**

It is strictly prohibited to create multivariate visualizations using variables that were not included in the previous section (univariate data analysis).

In [ ]:
# again, feel free to place as many cells to plot the visualizations,
# as well as describe to the main findings

# Final Plots (Effective Data Visualization)

In this section, you need to **enhance 3 multivariate visualizations** that were presented in the previous section of the report.
The goal is to enhance these visualizations so that they can be effectively presented to an audience unfamiliar with the dataset or with data analysis.
**Therefore, make sure that their size, colors, textures, and other visual elements are appropriate and convey the intended information to the audience.**

For your final plots, make sure you follow these steps:
1. Present the plot;
2. Provide a description of the visualization, highlighting the key findings that can be drawn from it.


**Hint**: take a look at the checklist based on Evergreen’s work to ensure your visualizations meet the best practices for clarity and impact.

In [ ]:
# your code goes here

# Digest

In this section, you should briefly summarize the main findings of your project. Additionally, provide a critical reflection on your work and the effort invested throughout the module. Highlight the aspects you believe were handled successfully, as well as those that, in hindsight, you would approach differently.

While this digest must contain at least 2,500 characters, it is intended solely for self-reflection purposes and will not be considered in the formal evaluation of your work.

```
Add your text here.
```

# Checkpoint Submission *(**Ignore this component in the final submission**)*

1. Save this report as a Jupyter Notebook (`.ipynb`), ensuring that all code cells are executed and their outputs are visible in the exported notebook;
2. Export a copy of the report as a PDF file (`.pdf`), ensuring that all code cells are executed and their outputs are visible in the exported document. If the generated PDF has formatting issues, include it in the submission anyway;
4. Copy the dataset (if smaller than 10 MB);
5. Compress all the files (the Jupyter Notebook, PDF, and dataset) into a single ZIP archive named `EquipeXX.zip`, replacing XX with your team number;
6. Upload the ZIP file to Canvas.

# Machine Learning (**post-checkpoint!**)

In this section, you must create at least **3 machine learning models** for the task at hand. Depending on the problem's nature, you must select from classification, regression, or clustering models.
It is also important that you:
* Select **an appropriate validation protocol**, providing a rationale for why it is appropriate for this specific task;
* Choose **a suitable set of evaluation metrics**, providing an explanation for each and describing how it contributes to evaluating the model's performance in the context of this specific task.

In [ ]:
# use as many cells as needed

# Final Steps (Submission)


1. Run all cells in the notebook and make sure every cell is executed successfully, with all outputs visible;
2. Save this report as a Jupyter Notebook (`.ipynb`);
3. Export a copy of the report as a PDF file (`.pdf`), ensuring that all code cells are executed and their outputs are visible in the exported document. If the generated PDF has formatting issues, include it in the submission anyway;
4. Copy the dataset (if smaller than 10 MB);
5. Compress all the files (the Jupyter Notebook, PDF, and dataset) into a single ZIP archive (`EquipeXX.zip`);
6. Upload the ZIP file to Canvas.